### DISCUSSION TITLE
Parser rule missed

### DISCUSSION THREAD

**Discussion by ZhangShushu123** [microsoft/gctoolkit#192] [Q&A] [answered]
I find gctoolkit can't parse my gc log and show that
```Missed: 2021-12-19T22:31:54.909+0800: 623.525: [Full GC (Allocation Failure) 2021-12-19T22:31:54.909+0800: 623.525: [CMS (concurrent mode failure): 2097152K->2097152K(2097152K), 9.4110998 secs] 3984639K->3984638K(3984640K), [Metaspace: 215219K->215219K(1241088K)], 9.4112098 secs] [Times: user=9.27 sys=0.15, real=9.41 secs]```
I think gctoolkit should add or modify a rule to match the line of gc log.
By the way, I suggest we should build an issue to record the missing rule which used to strengthen gctoolkit.

**Comment by karianna** [comment]
Converted this to an issue, it's a bug :-)

**Comment by ghost** [comment, marked as answer]
Hi @ZhangShushu123, Do you have the complete log for this missing line? If so, can you share it? If you would like the pleasure of fixing this bug I'm happy to assign it to you and I'm happy to assist.

This rule would be defined in CMSPatterns as a GCParseRule. It would be parsed and interpreted in GenerationalHeapParser. In this case I'd like to see the lines just before this line to confirm how to interrupt this. At the moment I'm seeing this as a Concurrent Mode Failure but as a Full GC caused by an allocation failure, not from a promotion failure. This feels like a large allocation from a mutator thread directly into tenured.

You can setup testing for the rule in ConcurrentMarkSweepParserRulesTest. There should be one other test that makes sure the capture groups match what you think they should match and then you may break the tests that are counting events.
